aspnet/mvc/controllers/filters.rst

```diff
@@ -1,8 +1,312 @@
+
+Adding this filter to a class or method is shown here:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Controllers/CachedController.cs
+  :language: c#
+  :emphasize-lines: 1-2,6
+  :lines: 7-14
+  :dedent: 4
+  
+.. _action-filters:
+
+Action Filters
+--------------
+*Action Filters* implement either the ``IActionFilter`` or ``IAsyncActionFilter`` interface and their execution surrounds the execution of action methods. Action filters are ideal for any logic that needs to see the results of model binding, or modify the controller or inputs to an action method. Additionally, action filters can view and directly modify the result of an action method.
+
+The ``OnActionExecuting`` method runs before the action method, so it can manipulate the inputs to the action by changing ``ActionExecutingContext.ActionArguments`` or manipulate the controller through ``ActionExecutingContext.Controller``. An ``OnActionExecuting`` method can short-circuit execution of the action method and subsequent action filters by setting ``ActionExecutingContext.Result``. Throwing an exception in an ``OnActionExecuting`` method will also prevent execution of the action method and subsequent filters, but will be treated as a failure instead of successful result.
+
+The ``OnActionExecuted`` method runs after the action method, and can see and manipulate the results of the action through the ``ActionExecutedContext.Result`` property. ``ActionExecutedContext.Canceled`` will be set to true if the action execution was short-circuited by another filter. ``ActionExecutedContext.Exception`` will be set to a non-null value if the action or a subsequent action filter threw an exception. Setting ``ActionExecutedContext.Exception`` to null effectively 'handles' an exception, and ``ActionExectedContext.Result`` will then be executed as if it were returned from the action method normally.
+
+For an ``IAsyncActionFilter`` the ``OnActionExecutionAsync`` combines all the possibilites of ``OnActionExecuting`` and ``OnActionExecuted``. A call to ``await next()`` on the ``ActionExecutionDelegate`` will execute any subsequent action filters and the action method, returning an ``ActionExecutedContext``. To short-circuit inside of an ``OnActionExecutionAsync``, set ``ActionExecutingContext.Result`` and do not call the ``ActionExectionDelegate``.
+
+The ``DurationActionFilter`` shown below uses a ``Stopwatch`` to measure the execution time of the action, for display on the page. The ``ViewData`` property of the ``ViewResult`` can be accessed and manipulated, including adding new values.
+
+.. _duration-action-filter:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/DurationActionFilter.cs
+  :language: c#
+  :emphasize-lines: 1,21-26
+  :lines: 8-35
+  :dedent: 4
+
+.. note:: If you just need to log how long an action takes to execute, rather than displaying it on the page, this is already done by the framework.
```

Comment:
You don't need to use this filter to log action times unless you need to display it on a page. To simple log action timing information use ...